File: apps/mobile/ios/Sources/FlagshipUI/Screens/KnockAuthorizeScreen.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// Visitor-side knock-authorize screen (docs/service-access-gating.md,
/// "Web-experience gating"). Reached from a `flagship://access?…` deeplink (the
/// box's knock page / QR) or a pasted "Process URL". Shows "Authorize this
/// site? <svc>.<server>" with Authorize / Cancel; Authorize AID-signs the
/// authorization behind the biometric gate (Keystore.deriveAccountId, like the
/// invite redeem) and POSTs it to the box. On success the browser session is
/// live and the owner is told to return to the website.
public struct KnockAuthorizeScreen: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(\.serviceAccessClient) private var client
    @Environment(\.securedSessionStore) private var store
    @Environment(ToastCenter.self) private var toasts

    private let serverDomain: String
    private let svc: String
    private let serviceRef: String
    private let pageId: String
    private let onDone: () -> Void

    @State private var vm: KnockAuthorizeViewModel?

    public init(
        serverDomain: String,
        svc: String,
        serviceRef: String,
        pageId: String,
        onDone: @escaping () -> Void = {}
    ) {
        self.serverDomain = serverDomain
        self.svc = svc
        self.serviceRef = serviceRef
        self.pageId = pageId
        self.onDone = onDone
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                switch vm?.phase {
                case .some(.authorized):
                    authorizedCard(c: c)
                default:
                    authorizeCard(c: c)
                }
                Spacer().frame(height: FS.space.s12)
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.top, FS.space.s4)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Authorize site")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            if vm == nil {
                vm = KnockAuthorizeViewModel(
                    client: client,
                    store: store,
                    serverDomain: serverDomain,
                    svc: svc,
                    serviceRef: serviceRef,
                    pageId: pageId)
            }
        }
    }

    /// The site label, e.g. `notes.home.alice.flagship.services` — derived the
    /// same way the VM persists it (drops the scheme for the headline).
    private var siteLabel: String {
        let url = SecuredSession.serviceUrl(svc: svc, serverDomain: serverDomain)
        return url.replacingOccurrences(of: "https://", with: "")
    }

    @ViewBuilder
    private func authorizeCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                Text("Authorize this site?")
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(c.text)
                Text(siteLabel)
                    .font(FS.font.mono())
                    .foregroundColor(c.text)
                    .accessibilityIdentifier("knock-authorize-site")
                Text("A browser is asking to open a restricted site on your Flagship cloud. Authorizing signs it in with your account so it can view the page; nothing else about you is shared.")
                    .font(FS.font.body())
                    .foregroundColor(c.textMuted)
                FSPrimaryButton(isAuthorizing ? "Authorizing…" : "Authorize", block: true, large: true) {
                    Task { await authorize() }
                }
                .disabled(isAuthorizing)
                .accessibilityIdentifier("knock-authorize-btn")
                FSSecondaryButton("Cancel", block: true) { onDone() }
                    .disabled(isAuthorizing)
                    .accessibilityIdentifier("knock-authorize-cancel")
                if case .some(.failed(let msg)) = vm?.phase {
                    Text(msg)
                        .font(FS.font.bodySm())
                        .foregroundColor(c.danger)
                        .accessibilityIdentifier("knock-authorize-error")
                }
            }
        }
    }

    @ViewBuilder
    private func authorizedCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                Label("Authorized", systemImage: "checkmark.seal.fill")
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(c.text)
                Text("Return to the website — it'll load now. You can refresh or stop this session anytime from Settings → Open secured sessions.")
                    .font(FS.font.body())
                    .foregroundColor(c.textMuted)
                FSPrimaryButton("Done", block: true, large: true) { onDone() }
                    .accessibilityIdentifier("knock-authorize-done")
            }
        }
    }

    private var isAuthorizing: Bool {
        if case .some(.authorizing) = vm?.phase { return true }
        return false
    }

    @MainActor
    private func authorize() async {
        guard let vm else { return }
        await vm.authorize()
        if case .failed(let msg) = vm.phase { toasts.error(msg) }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/KnockAuthorizeScreen.kt
// Web-experience gating authorizer screen (docs/service-access-gating.md,
// "Web-experience gating"). Reached from the knock page's
// flagship://access?... deep-link or a pasted "Get link" string. "Authorize"
// AID-signs the knock against the box (behind the biometric gate the other
// AID-signing flows use), then confirms — the browser's poll picks up the
// session and reloads into the content.

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.components.FSSecondaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.KnockAuthorizePhase
import com.flagshipserver.app.viewmodels.KnockAuthorizeViewModel
import kotlinx.coroutines.launch

/**
 * @param onDone invoked when the user dismisses the screen (Done / Cancel).
 */
@Composable
fun KnockAuthorizeScreen(
    serverId: String,
    svc: String,
    serviceRef: String,
    pageId: String,
    onDone: () -> Unit,
) {
    val scope = rememberCoroutineScope()
    val vm = remember(serverId, svc, serviceRef, pageId) {
        KnockAuthorizeViewModel(serverId = serverId, svc = svc, serviceRef = serviceRef, pageId = pageId)
    }
    val phase by vm.phase.collectAsState()

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(FS.space.s4),
        verticalArrangement = Arrangement.spacedBy(FS.space.s4),
    ) {
        when (val p = phase) {
            is KnockAuthorizePhase.Done -> FSCard(padding = PaddingValues(FS.space.s4)) {
                Column(verticalArrangement = Arrangement.spacedBy(FS.space.s3)) {
                    Text("Authorized", color = FS.colors.text, style = TextStyle(fontSize = 17.sp, fontWeight = FontWeight.SemiBold))
                    Text(
                        "Return to the website — it'll load now." +
                            if (p.browserAgent.isNotBlank()) "\n\nBrowser: ${p.browserAgent}" else "",
                        color = FS.colors.textMuted,
                        style = TextStyle(fontSize = 15.sp, lineHeight = 20.sp),
                    )
                    FSPrimaryButton(
                        label = "Done",
                        block = true,
                        large = true,
                        onClick = onDone,
                        modifier = Modifier.semantics { testTag = "knock-authorize-done" },
                    )
                }
            }
            else -> FSCard(padding = PaddingValues(FS.space.s4)) {
                Column(verticalArrangement = Arrangement.spacedBy(FS.space.s3)) {
                    Text("Authorize this site?", color = FS.colors.text, style = TextStyle(fontSize = 17.sp, fontWeight = FontWeight.SemiBold))
                    Text(
                        vm.target,
                        color = FS.colors.text,
                        style = TextStyle(fontSize = 15.sp, fontWeight = FontWeight.Medium),
                        modifier = Modifier.semantics { testTag = "knock-authorize-target" },
                    )
                    Text(
                        "A browser is asking to open this restricted site. Authorizing lets " +
                            "that browser in using your account — only do this if you started it.",
                        color = FS.colors.textMuted,
                        style = TextStyle(fontSize = 15.sp, lineHeight = 20.sp),
                    )
                    val busy = p is KnockAuthorizePhase.Authorizing
                    FSPrimaryButton(
                        label = if (busy) "Authorizing…" else "Authorize",
                        enabled = !busy,
                        block = true,
                        large = true,
                        onClick = { scope.launch { vm.authorize() } },
                        modifier = Modifier.semantics { testTag = "knock-authorize-accept" },
                    )
                    FSSecondaryButton(label = "Cancel", enabled = !busy, block = true, onClick = onDone)
                    (p as? KnockAuthorizePhase.Failed)?.let {
                        Text(
                            it.message,
                            color = FS.colors.danger,
                            style = TextStyle(fontSize = 14.sp),
                            modifier = Modifier.semantics { testTag = "knock-authorize-error" },
                        )
                    }
                }
            }
        }
        Spacer(Modifier.height(FS.space.s12))
    }
}
